File: docs/hermes-mcp-archive/hermes/modules/omniroute.py
"""
OmniRoute Client for Hermes MCP
Unified AI gateway client that replaces direct calls to Groq, OpenRouter, ZAI, Nous, etc.

All AI requests flow through OmniRoute's single endpoint with auto-fallback support.
"""

import os
import requests
import json
from typing import Optional, List, Dict, Any, Callable
from .prompts.hermes_system_prompt import HERMES_SYSTEM_PROMPT


class OmniRouteClient:
    """OmniRoute unified AI gateway client."""
    
    def __init__(self):
        self.base_url = (os.getenv('OMNIROUTE_BASE_URL', 'https://omniroute.ai')).rstrip('/')
        self.api_key = os.getenv('OMNIROUTE_API_KEY', '').strip()
        self.default_model = os.getenv('OMNIROUTE_MODEL', 'openai/gpt-4o-mini').strip()
        
        if not self.api_key:
            raise ValueError("OMNIROUTE_API_KEY environment variable is not set")
    
    def _get_headers(self) -> Dict[str, str]:
        """Get request headers with authorization."""
        return {
            'Authorization': f'Bearer {self.api_key}',
            'Content-Type': 'application/json',
        }
    
    def _build_request_body(
        self,
        prompt: str,
        model: str,
        system_prompt: str = "You are a helpful AI assistant.",
        json_mode: bool = False,
        messages: Optional[List[Dict[str, str]]] = None
    ) -> Dict[str, Any]:
        """Build the request body for OmniRoute API."""
        if messages is None:
            messages = [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": prompt}
            ]
        
        body = {
            "model": model,
            "messages": messages,
        }
        
        if json_mode:
            body["response_format"] = {"type": "json_object"}
        
        return body
    
    def call(
        self,
        prompt: str,
        model: Optional[str] = None,
        system_prompt: str = HERMES_SYSTEM_PROMPT,
        json_mode: bool = False,
        timeout: int = 60,
        fallback_models: Optional[List[str]] = None,
        messages: Optional[List[Dict[str, str]]] = None
    ) -> Dict[str, Any]:
        """
        Call OmniRoute with a prompt and optional parameters.
        
        Args:
            prompt: The user prompt/message
            model: Model override (default: OMNIROUTE_MODEL env or 'openai/gpt-4o-mini')
            system_prompt: System prompt override
            json_mode: Force JSON response mode
            timeout: Request timeout in seconds (default: 60)
            fallback_models: Models to try if primary fails
            messages: Pre-built messages array (overrides prompt)
        
        Returns:
            Dict with keys: reply, provider, model, error
        """
        if not prompt or not prompt.strip():
            return {
                'reply': '',
                'provider': None,
                'model': None,
                'error': 'Invalid prompt: must be a non-empty string',
            }
        
        model = model or self.default_model
        fallback_models = fallback_models or []
        models_to_try = [model] + [m for m in fallback_models if m]
        
        last_error = None
        
        for current_model in models_to_try:
            try:
                body = self._build_request_body(
                    prompt=prompt,
                    model=current_model,
                    system_prompt=system_prompt,
                    json_mode=json_mode,
                    messages=messages
                )
                
                response = requests.post(
                    f"{self.base_url}/v1/chat/completions",
                    headers=self._get_headers(),
                    json=body,
                    timeout=timeout
                )
                
                if not response.ok:
                    error_text = response.text
                    error_message = f"OmniRoute error: {response.status_code} {response.reason}"
                    
                    try:
                        error_json = response.json()
                        error_message += f" - {error_json.get('error', {}).get('message', error_text[:200])}"
                    except:
                        error_message += f" - {error_text[:200]}"
                    
                    raise Exception(error_message)
                
                data = response.json()
                raw_reply = data.get('choices', [{}])[0].get('message', {}).get('content', '')
                
                if not raw_reply:
                    raise Exception('OmniRoute returned empty response')
                
                return {
                    'reply': raw_reply,
                    'provider': 'omniroute',
                    'model': current_model,
                    'error': None,
                }
                
            except Exception as e:
                last_error = str(e)
                print(f"[OmniRoute] Model {current_model} failed: {last_error}")
                continue
        
        return {
            'reply': '',
            'provider': None,
            'model': None,
            'error': last_error or 'All OmniRoute models failed',
        }
    
    def call_with_messages(
        self,
        messages: List[Dict[str, str]],
        model: Optional[str] = None,
        json_mode: bool = False,
        timeout: int = 60,
        fallback_models: Optional[List[str]] = None
    ) -> Dict[str, Any]:
        """
        Call OmniRoute with a pre-built messages array.
        
        Args:
            messages: List of message dicts with 'role' and 'content' keys
            model: Model override
            json_mode: Force JSON response mode
            timeout: Request timeout in seconds
            fallback_models: Models to try if primary fails
        
        Returns:
            Dict with keys: reply, provider, model, error
        """
        if not messages:
            return {
                'reply': '',
                'provider': None,
                'model': None,
                'error': 'Invalid messages: must be a non-empty list',
            }
        
        model = model or self.default_model
        fallback_models = fallback_models or []
        models_to_try = [model] + [m for m in fallback_models if m]
        
        last_error = None
        
        for current_model in models_to_try:
            try:
                body = self._build_request_body(
                    prompt="",
                    model=current_model,
                    messages=messages,
                    json_mode=json_mode
                )
                
                response = requests.post(
                    f"{self.base_url}/v1/chat/completions",
                    headers=self._get_headers(),
                    json=body,
                    timeout=timeout
                )
                
                if not response.ok:
                    error_text = response.text
                    error_message = f"OmniRoute error: {response.status_code} {response.reason}"
                    
                    try:
                        error_json = response.json()
                        error_message += f" - {error_json.get('error', {}).get('message', error_text[:200])}"
                    except:
                        error_message += f" - {error_text[:200]}"
                    
                    raise Exception(error_message)
                
                data = response.json()
                raw_reply = data.get('choices', [{}])[0].get('message', {}).get('content', '')
                
                if not raw_reply:
                    raise Exception('OmniRoute returned empty response')
                
                return {
                    'reply': raw_reply,
                    'provider': 'omniroute',
                    'model': current_model,
                    'error': None,
                }
                
            except Exception as e:
                last_error = str(e)
                print(f"[OmniRoute] Model {current_model} failed: {last_error}")
                continue
        
        return {
            'reply': '',
            'provider': None,
            'model': None,
            'error': last_error or 'All OmniRoute models failed',
        }


# Singleton instance
_client: Optional[OmniRouteClient] = None


def get_client() -> OmniRouteClient:
    """Get or create the OmniRoute client singleton."""
    global _client
    if _client is None:
        _client = OmniRouteClient()
    return _client


def call_ai(
    prompt: str,
    model: Optional[str] = None,
    system_prompt: str = "You are a helpful AI assistant.",
    json_mode: bool = False,
    timeout: int = 60,
    fallback_models: Optional[List[str]] = None
) -> Dict[str, Any]:
    """
    Convenience function to call OmniRoute AI.
    
    Args:
        prompt: The user prompt/message
        model: Model override
        system_prompt: System prompt override
        json_mode: Force JSON response mode
        timeout: Request timeout in seconds
        fallback_models: Models to try if primary fails
    
    Returns:
        Dict with keys: reply, provider, model, error
    """
    client = get_client()
    return client.call(
        prompt=prompt,
        model=model,
        system_prompt=system_prompt,
        json_mode=json_mode,
        timeout=timeout,
        fallback_models=fallback_models
    )


def call_ai_with_messages(
    messages: List[Dict[str, str]],
    model: Optional[str] = None,
    json_mode: bool = False,
    timeout: int = 60,
    fallback_models: Optional[List[str]] = None
) -> Dict[str, Any]:
    """
    Convenience function to call OmniRoute AI with messages array.
    
    Args:
        messages: List of message dicts
        model: Model override
        json_mode: Force JSON response mode
        timeout: Request timeout in seconds
        fallback_models: Models to try if primary fails
    
    Returns:
        Dict with keys: reply, provider, model, error
    """
    client = get_client()
    return client.call_with_messages(
        messages=messages,
        model=model,
        json_mode=json_mode,
        timeout=timeout,
        fallback_models=fallback_models
    )

File: supabase/functions/_shared/omniroute.ts
/**
 * OmniRoute Client
 * Unified AI gateway client for DigitallyDefined OS
 *
 * OmniRoute ONLY — the single AI gateway. No direct provider calls,
 * no fallback models, no fallback providers.
 */

import { hermesSystemPrompt } from './hermesSystemPrompt.ts';

// Initialize AgentOps for monitoring
// deno-lint-ignore no-explicit-any
let agentops: any = null;
try {
  const AgentOps = await import('agentops');
  const AGENTOPS_API_KEY = Deno.env.get('AGENTOPS_API_KEY');
  if (AGENTOPS_API_KEY) {
    agentops = new AgentOps.default({ apiKey: AGENTOPS_API_KEY });
    console.log('✓ AgentOps initialized for OmniRoute');
  } else {
    console.log('⚠️  AgentOps API key not found - running without monitoring');
  }
} catch (e) {
  console.log('⚠️  AgentOps not available:', (e as Error)?.message || String(e));
}

// Normalize any configured base URL (with or without a trailing "/v1")
// into the canonical chat-completions endpoint.
export function omnirouteEndpoint(raw?: string | null): string {
  const base = (raw ?? Deno.env.get('OMNIROUTE_BASE_URL') ?? 'https://api.omniroute.ai/v1')
    .trim()
    .replace(/\/+$/, '')
    .replace(/\/v1$/, '');
  return `${base}/v1/chat/completions`;
}
const OMNIROUTE_API_KEY = (Deno.env.get('OMNIROUTE_API_KEY') || '').trim();
const DEFAULT_MODEL = (Deno.env.get('OMNIROUTE_MODEL') || 'auto').trim();
const DEFAULT_SYSTEM_PROMPT = hermesSystemPrompt;

/**
 * Call OmniRoute with a prompt and optional parameters
 *
 * @param {string} prompt - The user prompt/message
 * @param {object} options - Optional configuration
 * @param {string} options.model - Model override (default: OMNIROUTE_MODEL env or 'auto')
 * @param {string} options.systemPrompt - System prompt override
 * @param {boolean} options.jsonMode - Force JSON response mode
 * @param {number} options.timeout - Request timeout in ms (default: 60000)
 * @returns {Promise<{reply: string, provider: string, model: string, error: string|null}>}
 */
type OmniRouteOptions = {
  model?: string;
  systemPrompt?: string;
  jsonMode?: boolean;
  timeout?: number;
};

/**
 * Extract the assistant reply from either a standard JSON chat response or
 * an SSE stream body (this OmniRoute instance may stream even when
 * stream:false is requested, e.g. for the "auto" model).
 */
async function extractReply(response: Response): Promise<string> {
  const contentType = response.headers.get('content-type') || '';
  const text = await response.text();

  if (!contentType.includes('text/event-stream') && !text.trimStart().startsWith('data:')) {
    const data = JSON.parse(text);
    return data?.choices?.[0]?.message?.content || '';
  }

  let reply = '';
  for (const line of text.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed.startsWith('data:')) continue;
    const payload = trimmed.slice(5).trim();
    if (!payload || payload === '[DONE]') continue;
    try {
      const parsed = JSON.parse(payload);
      reply += parsed?.choices?.[0]?.delta?.content
        || parsed?.choices?.[0]?.message?.content
        || '';
    } catch {
      // Skip invalid JSON chunks
    }
  }
  return reply;
}

export async function omniRoute(prompt: string, options: OmniRouteOptions = {}) {
  if (!OMNIROUTE_API_KEY) {
    return {
      reply: '',
      provider: null,
      model: null,
      error: 'OMNIROUTE_API_KEY not configured in environment variables',
    };
  }

  if (!prompt || typeof prompt !== 'string' || !prompt.trim()) {
    return {
      reply: '',
      provider: null,
      model: null,
      error: 'Invalid prompt: must be a non-empty string',
    };
  }

  const model = options.model || DEFAULT_MODEL;
  const systemPrompt = options.systemPrompt || DEFAULT_SYSTEM_PROMPT;
  const jsonMode = options.jsonMode || false;
  const timeout = options.timeout || 60000;

  try {
    // AgentOps trace for LLM call
    const trace = agentops?.startTrace('omniroute_llm_call', {
      metadata: { model, systemPrompt: systemPrompt.slice(0, 100) }
    });

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    const requestBody: Record<string, unknown> = {
      model,
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: prompt.trim() },
      ],
    };

    // Add JSON mode if requested
    if (jsonMode) {
      requestBody.response_format = { type: 'json_object' };
    }

    const response = await fetch(omnirouteEndpoint(), {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${OMNIROUTE_API_KEY}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(requestBody),
      signal: controller.signal,
    });

    clearTimeout(timeoutId);

    if (!response.ok) {
      const errorText = await response.text();
      let errorMessage = `OmniRoute error: ${response.status} ${response.statusText}`;

      try {
        const errorJson = JSON.parse(errorText);
        errorMessage += ` - ${errorJson.error?.message || errorText.slice(0, 200)}`;
      } catch {
        errorMessage += ` - ${errorText.slice(0, 200)}`;
      }

      throw new Error(errorMessage);
    }

    const rawReply = await extractReply(response);

    if (!rawReply) {
      throw new Error('OmniRoute returned empty response');
    }

    // End trace on success
    trace?.end({ status: 'success', model });

    return {
      reply: rawReply,
      provider: 'omniroute',
      model,
      error: null,
    };
  } catch (error) {
    const lastError = (error as Error)?.message || String(error);
    console.error('[OmniRoute] Call failed:', lastError);

    // End trace on error
    try {
      agentops?.startTrace('omniroute_llm_error', {
        metadata: { model, error: lastError }
      })?.end({ status: 'error' });
    } catch (e) {
      // Ignore trace errors
    }

    // Single OmniRoute attempt — no fallback models, no provider switching.
    return {
      reply: '',
      provider: null,
      model: null,
      error: lastError,
    };
  }
}

/**
 * Call OmniRoute with streaming support (if available)
 * 
 * @param {string} prompt - The user prompt/message
 * @param {object} options - Optional configuration (same as omniRoute)
 * @param {function} onChunk - Callback for each streaming chunk
 * @returns {Promise<{reply: string, provider: string, model: string, error: string|null}>}
 */
export async function omniRouteStream(prompt: string, options: OmniRouteOptions = {}, onChunk?: (chunk: string, fullReply: string) => void) {
  if (!OMNIROUTE_API_KEY) {
    return {
      reply: '',
      provider: null,
      model: null,
      error: 'OMNIROUTE_API_KEY not configured in environment variables',
    };
  }

  if (!prompt || typeof prompt !== 'string' || !prompt.trim()) {
    return {
      reply: '',
      provider: null,
      model: null,
      error: 'Invalid prompt: must be a non-empty string',
    };
  }

  const model = options.model || DEFAULT_MODEL;
  const systemPrompt = options.systemPrompt || DEFAULT_SYSTEM_PROMPT;
  const jsonMode = options.jsonMode || false;
  const timeout = options.timeout || 60000;

  try {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    const requestBody: Record<string, unknown> = {
      model,
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: prompt.trim() },
      ],
      stream: true,
    };

    if (jsonMode) {
      requestBody.response_format = { type: 'json_object' };
    }

    const response = await fetch(omnirouteEndpoint(), {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${OMNIROUTE_API_KEY}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(requestBody),
      signal: controller.signal,
    });

    clearTimeout(timeoutId);

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`OmniRoute streaming error: ${response.status} ${response.statusText} - ${errorText.slice(0, 200)}`);
    }

    if (!response.body) {
      throw new Error('OmniRoute streaming response has no body');
    }
    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let fullReply = '';

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      const chunk = decoder.decode(value, { stream: true });
      const lines = chunk.split('\n').filter(line => line.trim() !== '');

      for (const line of lines) {
        if (line.startsWith('data: ')) {
          const data = line.slice(6);
          if (data === '[DONE]') continue;

          try {
            const parsed = JSON.parse(data);
            const content = parsed?.choices?.[0]?.delta?.content || '';
            if (content) {
              fullReply += content;
              if (typeof onChunk === 'function') {
                onChunk(content, fullReply);
              }
            }
          } catch {
            // Skip invalid JSON chunks
          }
        }
      }
    }

    return {
      reply: fullReply,
      provider: 'omniroute',
      model,
      error: null,
    };

  } catch (error) {
    return {
      reply: '',
      provider: null,
      model: null,
      error: (error as Error)?.message || String(error),
    };
  }
}

export default { omniRoute, omniRouteStream };